New load_csr() in apps.c simplifiying req, x509, and ca apps and allowing DER input to ca

DDvO opened the pull request:
New helper function `X509_REQ *load_csr(const char *file, int format, const char *desc)` 
in `apps/apps.c` simplifying and generalizing CSR loading in `req.c`, `x509.c`, and `ca.c`.
Moreover, new `-inform` and `-certform` CLI options to ca app now allowing DER input.
Tested with DER (rather than PEM) CSR input format in `80-test_ca.t`.
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

bernd-edlinger commented on apps/apps.c:
```diff
@@ -700,6 +700,29 @@ X509_CRL *load_crl(const char *infile, int format)
+    else if (format == FORMAT_PEM)
+        req = PEM_read_bio_X509_REQ(in, NULL, NULL, NULL);
+    else if (desc)
+        BIO_printf(bio_err, "unsupported format for loading %s\n", desc);
```
wouldn't it be nice to print which format was requested?
when this error message is printed, the one below is kind of duplicated.

DDvO commented:
@bernd-edlinger, thanks for your review and comments.

Indeed the output in case of unsupported format could be more helpful. I've now done this not only for `load_csr()` but also for other file loading functions, introducing helper functions `format2str() `and `print_format_error()`.

Doing so, I found that the use of the `desc `/ `descrip` string parameter was a bit inconsistent, with some functions having it and similar other functions not; so I added it where missing.
Moreover, when using this parameter, no checks had been done whether it is NULL, so I added such guards. This leads to the nice extra feature that error printing on file loading can be suppressed, which can be very handy for instance when trying to load files using differenct formats until (hopefully) this succeeds.

richsalz commented:
Nice work, @DDvO

bernd-edlinger commented on apps/apps.c:
```diff
@@ -686,18 +693,45 @@ X509_CRL *load_crl(const char *infile, int format)
-    if (x == NULL) {
-        BIO_printf(bio_err, "unable to load CRL\n");
+
+ end:
```
you should move this label after the `if (x == NULL && desc != NULL)`

DDvO replied:
This has already been done in my last commit of Friday: d918db0

bernd-edlinger commented on apps/apps.c:
```diff
@@ -686,18 +693,45 @@ X509_CRL *load_crl(const char *infile, int format)
+        goto end;
+    }
 
  end:
```
here as well.

DDvO replied:
Here I disagree: printing the format error is the fallback in case an unsupported format has been requested. This is not a general loading error, which is then reported after the "end:" label.

bernd-edlinger commented on apps/apps.c:
```diff
@@ -760,12 +794,19 @@ EVP_PKEY *load_key(const char *file, int format, int maybe_stdin,
+                           );
         goto end;
     }
  end:
```
and here?

DDvO replied:
Same as before.

bernd-edlinger replied:
No, I just see a goto end at the end of an else block, that goes to where the normal flow of
control would continue anyways, therefore I have to ask if that is really the intention.

DDvO replied:
Right, the "goto end" in this case is not really needed; I just left it there for symmetry with other error cases before. Yet since this may cause confusion I took it out, also in print_cert() and load_crl().

bernd-edlinger commented on apps/apps.c:
```diff
@@ -686,18 +693,45 @@ X509_CRL *load_crl(const char *infile, int format)
+    if (x == NULL && desc != NULL) {
+        BIO_printf(bio_err, "unable to load %s\n", desc);
         ERR_print_errors(bio_err);
         goto end;
```
What I meant with the comment above: doesn't that look like an endless loop?

DDvO replied:
Ah, this was a bad copy&paste mistake of mine, thanks. Removed.

DDvO commented:
Hmm, it looks like the Travis CI build https://travis-ci.org/openssl/openssl/jobs/321919199 failed due to some Debian package (update) issue?

FdaSilvaYY commented on apps/apps.c:
```diff
@@ -545,7 +545,7 @@ static int load_pkcs12(BIO *in, const char *desc,
     PKCS12 *p12;
     p12 = d2i_PKCS12_bio(in, NULL);
-    if (p12 == NULL) {
+    if (p12 == NULL && desc != NULL) {
```
A NULL pointer value in `desc ` should not shutdown error management, please ;) 
Better overwrite a NULL pointer to use a static default description like :
```
static const char desc_NA[] = "N/A";
...
if  (desc == NULL)
    desc = desc_NA
```

DDvO replied:
@FdaSilvaYY, did you see my above remark?

> Moreover, when using this parameter, no checks had been done whether it is NULL, so I added such guards. This leads to the nice extra feature that error printing on file loading can be suppressed, which can be very handy for instance when trying to load files using different formats until (hopefully) this succeeds.

So consider this a feature, not a bug ;)
Still I like the suggested technique for adding a default description.

I plan to provide a further PR that makes use of this possibility, for instance as follows:

    static const char desc_default[] = "certificate";

    if  (desc == NULL)
        desc = desc_default;
    ERR_set_mark();
    cert = load_cert(file, format, pass, NULL);
    if (cert == NULL && (format == FORMAT_PEM || format == FORMAT_ASN1)) {
        format = (format == FORMAT_PEM) ? FORMAT_ASN1 : FORMAT_PEM;
        cert = load_cert(file, format, pass, NULL);
    }
    if (cert == NULL) {
        ERR_clear_last_mark();
        ERR_print_errors(bio_err);
        BIO_printf(bio_err, "error: unable to load %s from '%s'\n", desc, file);
    } else {
        ERR_pop_to_mark();
    }

This would provide a convenient second try in case the given or default PEM/DER format does not fit. I also have an even more useful variant that determines the format for the second try from the file name extension (and in case this yields the same format, flips PEM vs. DER as given above).

What you do you think about this?

FdaSilvaYY commented on apps/apps.c:
```diff
@@ -615,23 +617,23 @@ static int load_cert_crl_http(const char *url, X509 **pcert, X509_CRL **pcrl)
-    if (rv != 1) {
-        BIO_printf(bio_err, "Error loading %s from %s\n",
-                   pcert ? "certificate" : "CRL", url);
+    if (rv != 1 && desc != NULL) {
```
Same `desc != NULL ` as above

FdaSilvaYY commented on apps/apps.c:
```diff
@@ -685,23 +691,45 @@ X509_CRL *load_crl(const char *infile, int format)
+        print_format_error(format, OPT_FMT_PEMDER);
+
+ end:
+    if (x == NULL && desc != NULL) {
```
Same `desc != NULL ` as above

FdaSilvaYY commented on apps/apps.c:
```diff
@@ -685,23 +691,45 @@ X509_CRL *load_crl(const char *infile, int format)
+        print_format_error(format, OPT_FMT_PEMDER);
 
  end:
+    if (req == NULL && desc != NULL) {
```
same `desc != NULL ` as above

FdaSilvaYY commented on apps/apps.c:
```diff
@@ -723,8 +751,8 @@ EVP_PKEY *load_key(const char *file, int format, int maybe_stdin,
             }
-            if (pkey == NULL) {
-                BIO_printf(bio_err, "cannot load %s from engine\n", key_descrip);
+            if (pkey == NULL && desc != NULL) {
```
Same `desc != NULL ` issue as above

FdaSilvaYY commented on apps/apps.c:
```diff
@@ -792,8 +827,8 @@ EVP_PKEY *load_pubkey(const char *file, int format, int maybe_stdin,
             pkey = ENGINE_load_public_key(e, file, ui_method, &cb_data);
-            if (pkey == NULL) {
-                BIO_printf(bio_err, "cannot load %s from engine\n", key_descrip);
+            if (pkey == NULL && desc != NULL) {
```
idem

FdaSilvaYY commented on apps/apps.c:
```diff
@@ -850,11 +885,19 @@ EVP_PKEY *load_pubkey(const char *file, int format, int maybe_stdin,
     BIO_free(key);
-    if (pkey == NULL)
-        BIO_printf(bio_err, "unable to load %s\n", key_descrip);
+    if (pkey == NULL && desc != NULL) {
```
idem

FdaSilvaYY commented on apps/apps.c:
```diff
@@ -933,9 +976,11 @@ static int load_certs_crls(const char *file, int format,
-        BIO_printf(bio_err, "unable to load %s\n",
-                   pcerts ? "certificates" : "CRLs");
-        ERR_print_errors(bio_err);
+        if (desc != NULL) {
```
idem

levitte commented:
At this point, I think we need to post-pone this PR to be post-1.1.1, as it's very much an added features thing.

Also, I'm thinking that this kind of change should really make use of the `OSSL_STORE` API.

DDvO commented:
I've meanwhile rebased also this PR such that we can hopefully integrate it with #11470 soon.
Travis CI went fine (while the AppVeyor runs are still queued, but I don't expect problems there).
Anything concrete that I can quickly improve?

t8m commented on apps/x509.c:
```diff
@@ -568,14 +568,8 @@ int x509_main(int argc, char **argv)
-        req = PEM_read_bio_X509_REQ(in, NULL, NULL, NULL);
-        BIO_free(in);
 
+        req = load_csr(infile, informat, "certificate request input");
```
The ERR_print_errors() below seems to be superfluous as you're already calling it in load_csr().

DDvO replied:
Thanks @t8m for your comment.
I've cleaned up the use of `ERR_print_errors()` also in four other apps and in `apps/lib/apps.c` (and made sure that all error message there consistently begin upper-case).

t8m commented:
Hopefully nobody depends on exact wording (and case) for the error messages in scripts.

Could you please amend the commit message of the last commit so the subject is shorter and you add description of the changes in message body.

DDvO commented:
> Hopefully nobody depends on exact wording (and case) for the error messages in scripts.

It would be pretty unwise to do so. 
This is not the first time error messages are adapted.

> Could you please amend the commit message of the last commit so the subject is shorter and you add description of the changes in message body.

Sure - done.
I anyway plan to partly squash and reword the commit messages ones the changes are stable.

t8m commented:
I think if CI passes this could be approved unless you plan further changes.

DDvO commented:
Pleased to hear. I don't plan to do further changes unless request by a review.
Then let's just wait for the CI checks and in the meantime I'll prepare the cleanup for the commit structure and comments...

DDvO commented:
Since Travis CI went fine I've already updated the PR to the final commit structure I propose.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but this PR has failing CI tests.  Once the tests pass it will get moved to 'approval: ready to merge' automatically, alternatively please review and set the label manually.

DDvO commented:
For some reason AppVeyor is still queued on this PR (and others as well) since about two days,
and as far as I recall it already went fine earlier on this PR.
So I've just manually marked this "ready to merge" and will do the merge now...

DDvO commented:
Merged - thanks @t8m for your swift review!
